### PR TITLE
Added connection type reference

### PR DESCRIPTION
**Pull request by guyblink**
Can define a connection type reference in the plugin.yaml and add additional attributes on top of it.

`provider: "blink"
connection_types:
  aws-extended:
    fields:
        test:
          field_type: "input"
          required: true
          placeholder: "Test"
          input_type: "text"
    reference: "aws"`

### CODE REVIEW

File: plugin/proto/plugin.proto [hunk 1]
```diff
@@ -28,6 +28,7 @@ message PluginDescription {
   string image = 6;
   map<string, Connection> connections = 7;
   string version = 8;
+  string connection_reference = 9;
```
Comment (danielsblinkops) [inline plugin/proto/plugin.proto:31]: Why?

File: plugin/connections/connection.go [hunk 2]
```diff
@@ -22,12 +22,14 @@ type ConnectionField struct {
 // For each requested connection an instance of it will be sent
 // for every action executed.
 type Connection struct {
-	Name   string
-	Fields map[string]ConnectionField
+	Name      string
+	Fields    map[string]ConnectionField
+	Reference string
```
Comment (danielsblinkops) [inline plugin/connections/connection.go:27]: Do we want the connection to both reference and define new fields?

**Review by danielsblinkops — CHANGES_REQUESTED** [review]

File: plugin/connections/connection.go (hunk 2 above)
Comment (guyblink) [reply plugin/connections/connection.go:27]: Yup. It adds them on top of the referenced fields

File: plugin/proto/plugin.proto (hunk 1 above)
Comment (guyblink) [reply plugin/proto/plugin.proto:31]: This is the base connection type we reference

Comment (danielsblinkops) [reply plugin/proto/plugin.proto:31]: Each connection should have a base reference type, not the whole plugin description

Comment (guyblink) [reply plugin/proto/plugin.proto:31]: Reference is now map[string]string -> new-type-name: referenced-type

File: plugin/proto/plugin.proto [hunk 3]
```diff
@@ -28,6 +28,7 @@ message PluginDescription {
   string image = 6;
   map<string, Connection> connections = 7;
   string version = 8;
+  map<string, string> connection_references = 9;
```
Comment (danielsblinkops) [inline plugin/proto/plugin.proto:31]: But this data is inside the connection map
map<string, Connection> connections = 7
Why duplicate it?

**Review by danielsblinkops — CHANGES_REQUESTED** [review]

File: plugin/plugin.go [hunk 4]
```diff
@@ -38,12 +38,13 @@ type Output struct {
...
-	Tags        []string                          `yaml:"tags"`
-	Provider    string                            `yaml:"provider"`
-	Connections map[string]connections.Connection `yaml:"connections"`
-	Version     string                            `yaml:"version"`
+	Name                  string                            `yaml:"name"`
+	Description           string                            `yaml:"description"`
+	Tags                  []string                          `yaml:"tags"`
+	Provider              string                            `yaml:"provider"`
+	Connections           map[string]connections.Connection `yaml:"connection_types"`
+	ConnectionsReferences map[string]string                 `yaml:"connection_type_references"`
```
Comment (danielsblinkops) [inline plugin/plugin.go:46]: Why do you need the map if Connection already holds a 'references' field?

**Review by danielsblinkops — CHANGES_REQUESTED** [review]

File: plugin/proto/plugin.proto (hunk 3 above)
Comment (guyblink) [reply plugin/proto/plugin.proto:31]: It's not there, references are a different field, see the PR description for example. After I parse the connection type reference map, I add each reference to its connection type
Connection struct {
	Name      string
	Fields    map[string]ConnectionField
	Reference string
}

File: plugin/plugin.go (hunk 4 above)
Comment (guyblink) [reply plugin/plugin.go:46]: In the plugin yaml you have references and connection_types separately. I later on merge each connection with its reference. Makes it clearer when defining in yaml.

**Review by itsofirblink — APPROVED** [review]

**Pull request merged**